File: cpp/interview/classic/rwlock/rwlock.cpp
// 读写锁（读者优先）：与 Python 版同构（C++17）
#include <bits/stdc++.h>
#include <cassert>
#include <chrono>
#include <mutex>
#include <thread>
using namespace std;

class RWLock {
public:
    void read_lock() {
        lock_guard<mutex> lk(guard_);
        if (++readers_ == 1) write_.lock();
    }

    void read_unlock() {
        lock_guard<mutex> lk(guard_);
        if (--readers_ == 0) write_.unlock();
    }

    void write_lock() { write_.lock(); }

    void write_unlock() { write_.unlock(); }

private:
    mutex guard_;
    mutex write_;
    int readers_ = 0;
};

int main() {
    RWLock rw;
    atomic<int> x{0};
    atomic<bool> go{false};

    auto reader = [&] {
        while (!go.load()) this_thread::yield();
        for (int i = 0; i < 20; ++i) {
            rw.read_lock();
            int v = x.load();
            (void)v;
            rw.read_unlock();
            this_thread::sleep_for(chrono::microseconds(10));
        }
    };

    auto writer = [&] {
        while (!go.load()) this_thread::yield();
        for (int i = 0; i < 50; ++i) {
            rw.write_lock();
            x.fetch_add(1);
            rw.write_unlock();
            this_thread::sleep_for(chrono::microseconds(5));
        }
    };

    vector<thread> ts;
    for (int i = 0; i < 4; ++i) ts.emplace_back(reader);
    ts.emplace_back(writer);
    go = true;
    for (auto& t : ts) t.join();
    assert(x.load() == 50);
    cout << "rwlock OK" << endl;
    return 0;
}

File: python/interview/classic/rwlock/rwlock.py
"""读写锁（读者优先）：读者共享 `readers` 计数，首个读者持写互斥、末读者释放。"""

from __future__ import annotations

import threading
import time
from contextlib import contextmanager
from typing import Generator


class RWLock:
    """多读者并发读；写者独占。写者可能饥饿（读者优先）。"""

    def __init__(self) -> None:
        self._guard = threading.Lock()
        self._write = threading.Lock()
        self._readers = 0

    @contextmanager
    def read_lock(self) -> Generator[None, None, None]:
        with self._guard:
            self._readers += 1
            if self._readers == 1:
                self._write.acquire()
        try:
            yield
        finally:
            with self._guard:
                self._readers -= 1
                if self._readers == 0:
                    self._write.release()

    @contextmanager
    def write_lock(self) -> Generator[None, None, None]:
        self._write.acquire()
        try:
            yield
        finally:
            self._write.release()


if __name__ == "__main__":
    rw = RWLock()
    shared = {"x": 0}
    barrier = threading.Barrier(5)

    def reader() -> None:
        barrier.wait()
        for _ in range(20):
            with rw.read_lock():
                _ = shared["x"]
            time.sleep(0.001)

    def writer() -> None:
        barrier.wait()
        for _ in range(50):
            with rw.write_lock():
                shared["x"] = shared["x"] + 1
            time.sleep(0.0005)

    ts = [threading.Thread(target=reader) for _ in range(4)]
    ts.append(threading.Thread(target=writer))
    for t in ts:
        t.start()
    for t in ts:
        t.join()
    assert shared["x"] == 50
    print("rwlock OK")
